Governance.md

```diff
@@ -0,0 +1,78 @@
+# Governance of the rules_scala
+
+These bazel rules for scala are under the bazelbuild org but governed by a
+contributors from the community.
+
+## Goals of this document
+
+Clarify:
+0. values of the project.
+1. how to get a PR merged
+2. how long you should need to wait for a review
+3. how to resolve conflicts in project direction
+4. how to make larger and disruptive changes
+
+## Values of rules_scala
+
+1. Correctness: bazel values fully reproducible builds which are reliable and hermetic (only depend on the current state of your repository)
+2. Speed: we want to build as fast as possible. In a large repository, we believe caching and remote parallel builds are the only way to achieve this. Without correctness, we can't cache and use remote builds, so correctness enables speed. We will almost never trade off correctness for speed.
+3. Usability: we want it to be as easy as possible to adopt and to maintain a build with these rules. Usability is a lower priority than correctness or speed. We will virtually never trade either correctness or speed for more usability. Usability includes stability: we seek to absolutely minimize breaking changes. We will prefer to keep some warts or deprecated methods than force users to constantly make changes to keep their builds on the latest version of the rules.
+
```

Comment:
Do you think "configurability" or "extensibility" fits here somewhere? The flexibility that bazel gives us to easily add new rules and functionality seems like one of its super powers.

Reply:
Let's discuss. I think configuration is generally bad. Configuration in my opinion is often punting work to users. I think ideally, there should be one and only way way to do things, and configuration happens in unfortunate situations when that falls over.

So, I would certainly not want "every bell and whistle can be toggled in every setting" as a value, as you can see, I want quite the opposite.

We do have to support configuration sometimes when we can't avoid it because there is no agreement on how things should be done. But when there is say 95% agreement, or 99% agreement, making something an option for a corner case puts cost on everyone. Keep in mind, each config option can double our needed testing space: adding many options exponentially (in the technical sense) expands our needs for testing. This is a cost on all maintainers and contributors.

Reply:
If there is only one way to do things, then the rule set will likely be very opinionated and the user base likely very homogeneous. I also agree that having oodles of switches makes setting up the tool challenging.

Modularity and configurability are one of Bazel's strengths. I would rather we have high quality documentation to help people get set up with good defaults, and then provide a modular architecture for the rules, so users can customize them to their use case. We keep testing primarily around the opinionated default and put warning signs in the docs for when people do non-standard things, but we still support their use cases.

Reply:
Generally, I agree that configuration is bad _if_ you don't provide guidance or an opinion and leave the door wide open. Opinionated defaults provides the best of both worlds to me: an out-of-the-box that works for most folks, and an out in case someone really needs it.

Reply:
Generally, I agree that configuration is bad if you don't provide guidance or an opinion and leave the door wide open. Opinionated defaults provides the best of both worlds to me: an out-of-the-box that works for most folks, and an out in case someone really needs it.

Reply:
I'm wary for adding configurability as a value. This is another thing I would punt on for 2-3 months.

Reply:
I'm concerned that if we do not reach consensus around configurability, then we're going probably going to run into disagreements multiple times in PRs over the next few months as we join efforts - a strongly opinionated ruleset works well when everyone agrees on that opinion. Configuration enables differing opinions to coexist in the same ruleset.

Reply:
It also can lead to confused users and harder to maintain codebase.
To be clear- I'm not saying we should have just one way for doing things but at the same time I'm not sold it's a value of its own.

Reply:
Usability: We want it to be as easy as possible to adopt and to maintain a build with these rules. However, in the pursuit of usability, we cannot sacrifice correctness or speed. We will seek to implement the best tools and methods and avoid shortcuts when creating first class support in Bazel for Scala.
Stability: We seek to minimize breaking changes to those that are absolutely necessary. We will avoid forcing users to constantly make changes to keep their builds on the latest version of the rules.